File: src/rbs_rag/retrieval.py
from __future__ import annotations

import logging
from dataclasses import dataclass

from .embeddings import EmbeddingProvider
from .models import Chunk, SearchResult
from .reranking import create_reranker
from .store import SQLiteRagStore
from .text import bm25_scores, cosine_similarity
from .vector_store import QdrantVectorStore

log = logging.getLogger(__name__)


@dataclass(slots=True)
class HybridRetriever:
    store: SQLiteRagStore
    embedding_provider: EmbeddingProvider
    tenant_id: str
    top_k: int = 20
    rerank_top_k: int = 8
    final_context_k: int = 5
    dense_weight: float = 0.55
    sparse_weight: float = 0.45
    vector_store: QdrantVectorStore | None = None
    collection: str = "rag_chunks"

    def search(self, query: str, knowledge_base_id: str, filters: dict[str, str] | None = None) -> list[SearchResult]:
        results, _ = self.search_with_profile(query, knowledge_base_id, filters)
        return results

    def search_with_profile(self, query: str, knowledge_base_id: str, filters: dict[str, str] | None = None) -> tuple[list[SearchResult], dict[str, float]]:
        import time
        t0 = time.perf_counter()

        query_embedding = self.embedding_provider.embed([query])[0]
        t_emb_ms = (time.perf_counter() - t0) * 1000.0

        qdrant_results: list[tuple[Chunk, float]] = []

        # Phase 1: Qdrant vector search
        t_qd0 = time.perf_counter()
        if self.vector_store:
            try:
                kb_filter = {"knowledge_base_id": knowledge_base_id}
                if filters:
                    kb_filter.update(filters)
                qdrant_chunks = self.vector_store.search(
                    self.collection,
                    query_embedding,
                    top_k=self.top_k * 2,
                    filters=kb_filter,
                )
                qdrant_results = [(c, c.metadata.get("_score", 0.0)) for c in qdrant_chunks]
            except Exception as exc:
                log.warning("Qdrant search failed, falling back to SQLite: %s", exc)
        t_qd_ms = (time.perf_counter() - t_qd0) * 1000.0

        # Phase 2: fallback SQLite brute-force if Qdrant returned nothing
        t_db0 = time.perf_counter()
        sqlite_chunks = self.store.list_chunks(self.tenant_id, knowledge_base_id, filters) if not qdrant_results else []
        t_db_ms = (time.perf_counter() - t_db0) * 1000.0

        all_chunks: list[Chunk] = []
        if qdrant_results:
            all_chunks = [c for c, _ in qdrant_results]
        elif sqlite_chunks:
            all_chunks = sqlite_chunks
        else:
            return [], {"db_fetch_ms": round(t_db_ms + t_qd_ms, 2), "dense_emb_ms": 0.0, "vector_sim_ms": 0.0, "bm25_ms": 0.0, "rerank_ms": 0.0, "total_retrieval_ms": round((time.perf_counter() - t0) * 1000.0, 2), "chunks_scanned": 0}

        # Phase 3: Compute scores (only needed for SQLite fallback or hybrid fusion)
        t_sim0 = time.perf_counter()
        dense_scores = {}
        if not qdrant_results:
            for chunk in all_chunks:
                dense_scores[chunk.chunk_id] = cosine_similarity(query_embedding, chunk.embedding)
        else:
            for chunk, score in qdrant_results:
                dense_scores[chunk.chunk_id] = score
        t_sim_ms = (time.perf_counter() - t_sim0) * 1000.0

        # Phase 4: Sparse BM25
        t_bm25_0 = time.perf_counter()
        sparse_scores = bm25_scores(query, {chunk.chunk_id: chunk.text for chunk in all_chunks})
        t_bm25_ms = (time.perf_counter() - t_bm25_0) * 1000.0

        # Phase 5: Fusion
        t_rerank0 = time.perf_counter()
        dense_rank = _rank_scores(dense_scores)
        sparse_rank = _rank_scores(sparse_scores)

        results = []
        for chunk in all_chunks:
            dense = dense_scores.get(chunk.chunk_id, 0.0)
            sparse = sparse_scores.get(chunk.chunk_id, 0.0)
            fused = self.dense_weight * dense_rank.get(chunk.chunk_id, 0.0) + self.sparse_weight * sparse_rank.get(chunk.chunk_id, 0.0)
            results.append(SearchResult(chunk=chunk, score=fused, dense_score=dense, sparse_score=sparse))

        candidates = sorted(results, key=lambda item: item.score, reverse=True)[:self.top_k]

        reranker = create_reranker("local")
        final_results = reranker.rerank(query, candidates[:self.rerank_top_k], self.final_context_k)
        t_rerank_ms = (time.perf_counter() - t_rerank0) * 1000.0

        total_retrieval_ms = (time.perf_counter() - t0) * 1000.0

        profile = {
            "db_fetch_ms": round(t_db_ms + t_qd_ms, 2),
            "dense_emb_ms": round(t_emb_ms, 2),
            "vector_sim_ms": round(t_sim_ms, 2),
            "bm25_ms": round(t_bm25_ms, 2),
            "rerank_ms": round(t_rerank_ms, 2),
            "total_retrieval_ms": round(total_retrieval_ms, 2),
            "chunks_scanned": len(all_chunks),
            "qdrant_hit": bool(qdrant_results),
        }

        return final_results, profile


def _rank_scores(scores: dict[str, float]) -> dict[str, float]:
    ordered = sorted(scores.items(), key=lambda item: item[1], reverse=True)
    ranked: dict[str, float] = {}
    for rank, (chunk_id, score) in enumerate(ordered, start=1):
        if score <= 0:
            ranked[chunk_id] = 0.0
        else:
            ranked[chunk_id] = 1.0 / rank
    return ranked

File: rbs-rag-node/src/retrieval.ts
import { EmbeddingProvider } from './embeddings.js';
import { Chunk, SearchResult } from './types/models.js';
import { createReranker, Reranker } from './reranking.js';
import { DbStore } from './store.js';
import { QdrantVectorStore } from './vectorStore.js';
import { bm25Scores, normalizeQuery } from './text.js';

interface RetrievalProfile {
  dbFetchMs: number;
  denseEmbMs: number;
  vectorSimMs: number;
  bm25Ms: number;
  rerankMs: number;
  totalRetrievalMs: number;
  chunksScanned: number;
  qdrantHit: boolean;
}

export class HybridRetriever {
  private store: DbStore;
  private embeddingProvider: EmbeddingProvider;
  private vectorStore: QdrantVectorStore | null;
  private reranker: Reranker;
  private config: {
    topK: number;
    rerankTopK: number;
    finalContextK: number;
    denseWeight: number;
    sparseWeight: number;
  };

  constructor(
    store: DbStore,
    embeddingProvider: EmbeddingProvider,
    vectorStore: QdrantVectorStore | null,
    config: {
      topK: number;
      rerankTopK: number;
      finalContextK: number;
      denseWeight: number;
      sparseWeight: number;
      reranker: string;
    }
  ) {
    this.store = store;
    this.embeddingProvider = embeddingProvider;
    this.vectorStore = vectorStore;
    this.reranker = createReranker(config.reranker);
    this.config = config;
  }

  async search(
    query: string,
    knowledgeBaseId: string,
    tenantId: string,
    filters?: Record<string, string>
  ): Promise<{ results: SearchResult[]; profile: RetrievalProfile }> {
    const t0 = performance.now();

    // Query Normalization
    const normalizedQuery = normalizeQuery(query);

    const qEmbedding = (await this.embeddingProvider.embed([normalizedQuery]))[0];
    const tEmb = (performance.now() - t0);

    // 1. Fetch ALL chunks from SQLite for this tenant & KB to act as the full candidate pool
    const tDb0 = performance.now();
    const allChunks = await this.store.listChunks(tenantId, knowledgeBaseId, filters);
    const tDb = (performance.now() - tDb0);

    // 2. Query Qdrant for top dense hits
    let qdrantChunks: Chunk[] = [];
    const tQd0 = performance.now();
    let qdrantHit = false;
    if (this.vectorStore?.initialized) {
      try {
        const qFilter: Record<string, string> = { knowledge_base_id: knowledgeBaseId };
        if (filters) Object.assign(qFilter, filters);
        qdrantChunks = await this.vectorStore.search('rag_chunks', qEmbedding, this.config.topK * 2, qFilter);
        qdrantHit = qdrantChunks.length > 0;
      } catch (err) {
        console.warn('Qdrant search failed:', (err as Error).message);
      }
    }
    const tQd = (performance.now() - tQd0);

    // 3. Exclude non-searchable document types (images, CSS, JS, HTML boilerplate, empty output)
    const EXCLUDED_DOC_TYPES = new Set(['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'svg', 'css', 'js', 'html']);
    
    const filteredChunks = allChunks.filter(chunk => {
      if (!chunk.text || !chunk.text.trim()) return false;
      const docType = (chunk.metadata?.document_type || '').toLowerCase();
      const docName = (chunk.metadata?.document_name || '').toLowerCase();
      const docExt = docName.split('.').pop() || '';
      return !EXCLUDED_DOC_TYPES.has(docType) && !EXCLUDED_DOC_TYPES.has(docExt);
    });

    const filteredQdrantChunks = qdrantChunks.filter(chunk => {
      if (!chunk.text || !chunk.text.trim()) return false;
      const docType = (chunk.metadata?.document_type || '').toLowerCase();
      const docName = (chunk.metadata?.document_name || '').toLowerCase();
      const docExt = docName.split('.').pop() || '';
      return !EXCLUDED_DOC_TYPES.has(docType) && !EXCLUDED_DOC_TYPES.has(docExt);
    });

    if (!filteredChunks.length) {
      return {
        results: [],
        profile: {
          dbFetchMs: tDb,
          denseEmbMs: tEmb,
          vectorSimMs: tQd,
          bm25Ms: 0,
          rerankMs: 0,
          totalRetrievalMs: (performance.now() - t0),
          chunksScanned: 0,
          qdrantHit,
        },
      };
    }

    // 4. Compute dense scores
    const tSim0 = performance.now();
    const denseScores: Record<string, number> = {};
    for (const chunk of filteredChunks) {
      denseScores[chunk.chunkId] = 0;
    }

    if (qdrantHit) {
      for (const chunk of filteredQdrantChunks) {
        denseScores[chunk.chunkId] = chunk.metadata._score || 0;
      }
    } else {
      // cosine similarity fallback
      for (const chunk of filteredChunks) {
        if (chunk.embedding.length) {
          denseScores[chunk.chunkId] = cosineSimilarity(qEmbedding, chunk.embedding);
        }
      }
    }
    const tSim = (performance.now() - tSim0);

    // 5. BM25 on the full set of filtered candidates
    // TODO: Replace in-memory BM25 scan with an inverted-index based implementation for large corpora.
    const tBm25 = performance.now();
    const docTexts: Record<string, string> = {};
    for (const chunk of filteredChunks) docTexts[chunk.chunkId] = chunk.text;
    const sparseScores = bm25Scores(normalizedQuery, docTexts);
    const tBm25End = (performance.now() - tBm25);

    // 6. Reciprocal Rank Fusion (RRF k=60)
    const tRerank0 = performance.now();
    const denseRank = rankScores(denseScores, 60);
    const sparseRank = rankScores(sparseScores, 60);

    const results: SearchResult[] = filteredChunks.map(chunk => ({
      chunk,
      score: 0,
      denseScore: denseScores[chunk.chunkId] || 0,
      sparseScore: sparseScores[chunk.chunkId] || 0,
      rerankScore: 0,
    }));

    for (const r of results) {
      const d = denseRank[r.chunk.chunkId] || 0;
      const s = sparseRank[r.chunk.chunkId] || 0;
      r.score = this.config.denseWeight * d + this.config.sparseWeight * s;
    }

    // Slicing top candidates (topK = 20) before feeding to reranker (Top 20 candidates only)
    const candidates = results.sort((a, b) => b.score - a.score).slice(0, this.config.topK);
    const finalResults = this.reranker.rerank(normalizedQuery, candidates.slice(0, this.config.rerankTopK), this.config.finalContextK);
    const tRerank = (performance.now() - tRerank0);

    return {
      results: finalResults,
      profile: {
        dbFetchMs: tDb,
        denseEmbMs: tEmb,
        vectorSimMs: qdrantHit ? tQd : tSim,
        bm25Ms: tBm25End,
        rerankMs: tRerank,
        totalRetrievalMs: (performance.now() - t0),
        chunksScanned: filteredChunks.length,
        qdrantHit,
      },
    };
  }
}

function rankScores(scores: Record<string, number>, k = 60): Record<string, number> {
  const ordered = Object.entries(scores).sort((a, b) => b[1] - a[1]);
  const ranked: Record<string, number> = {};
  for (let i = 0; i < ordered.length; i++) {
    const [id, score] = ordered[i];
    ranked[id] = score <= 0 ? 0 : 1 / (k + i + 1);
  }
  return ranked;
}

function cosineSimilarity(a: number[], b: number[]): number {
  if (!a.length || !b.length || a.length !== b.length) return 0;
  let dot = 0, na = 0, nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  na = Math.sqrt(na);
  nb = Math.sqrt(nb);
  if (na === 0 || nb === 0) return 0;
  return dot / (na * nb);
}
